docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,958 @@
+- Transition to new, improved, K8s API conventions aligned conditions
+
+Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
+Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.
+
+```golang
+type MachineSetStatus struct {
+
+    // The number of ready replicas for this MachineSet. A machine is considered ready when Machine's Ready condition is true.
+    // +optional
+    ReadyReplicas int32 `json:"readyReplicas"`
+	
+	// The number of available replicas for this MachineSet. A machine is considered available when Machine's Available condition is true.
+	// +optional
+	AvailableReplicas int32 `json:"availableReplicas"`
+
+    // The number of up-to-date replicas for this MachineSet. A machine is considered up-to-date when Machine's UpToDate condition is true.
+    // +optional
+    UpToDateReplicas int32 `json:"upToDateReplicas"`
+
+    // Represents the observations of a MachineSet's current state.
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+    // Other fields...
+    // NOTE: `FailureReason`, `FailureMessage` fields won't be there anymore
+}
+```
+
+| v1beta1 (current)                    | v1beta2 (tentative Q1 2025)                                 | v1beta1 removal (tentative Q1 2026) |
+|--------------------------------------|-------------------------------------------------------------|-------------------------------------|
+| `ExprimentalReadyReplicas` (new)     | `ReadyReplicas` (renamed)                                   | `ReadyReplicas`                     |
+| `ExprimentalAvailableReplicas` (new) | `AvailableReplicas` (renamed)                               | `AvailableReplicas`                 |
+|                                      | `BackCompatibilty` (new)                                    | (removed)                           |
+| `ReadyReplicas` (deprecated)         | `BackCompatibilty.ReadyReplicas` (renamed) (deprecated)     | (removed)                           |
+| `AvailableReplicas` (deprecated)     | `BackCompatibilty.AvailableReplicas` (renamed) (deprecated) | (removed)                           |
+| `FailureReason` (deprecated)         | `BackCompatibilty.FailureReason` (renamed) (deprecated)     | (removed)                           |
+| `FailureMessage` (deprecated)        | `BackCompatibilty.FailureMessage` (renamed) (deprecated)    | (removed)                           |
+| `Conditions`                         | `BackCompatibilty.Conditions` (renamed) (deprecated)        | (removed)                           |
+| `ExperimentalConditions` (new)       | `Conditions` (renamed)                                      | `Conditions`                        |
+| `UpToDateReplicas` (new)             | `UpToDateReplicas`                                          | `UpToDateReplicas`                  |
```

Comment:
Why is 'UpToDateReplicas' not prefixed with 'Experimental' yet it is also new?

Reply:
The reason why we have to use experimental is that we have naming conflicts. E.g. with v1beta2 we want to have the same condition field with a different type. To be able to already introduce something now and allow a smooth migration we need a different name for v1beta1.

We only did this "name migration" for fields with conflicts. `UpToDateReplicas` is an entirely new field so there is no conflict to handle

Reply:
resolving this in favour of https://github.com/kubernetes-sigs/cluster-api/pull/10897/files#r1688511927 where we are discussing moving all experimental/new fields under `v1Beta2` (also trying to keep the number of open threads under control 😅 )